### PR TITLE
fix: convert `run_pipeline_no_finalize` from recursive to iterative in order to avoid stack overflows

### PR DESCRIPTION
**Pull request by LukeMathWalker**
I ran into a stack overflow using `polars` in one of our projects—I was able to pin it down to `run_pipeline_no_finalize`, which is indeed recursive.

image

I've converted the algorithm to be iterative, which was not super-straightforward. Happy to iterate (pun intended) on the code to clean it up however you think it's best.

### CODE REVIEW

File: crates/polars-pipe/src/pipeline/dispatcher.rs [hunk 1]
```diff
@@ -330,89 +331,6 @@ impl PipeLine {
...
-                        ec,
-                        operator_start,
-                        operator_end,
-                        src,
-                    )?;
-                    next_batches = next_batches2;
-
-                    if let Some(SinkResult::Finished) = sink_result {
-                        sink_finished = true;
-                        break;
```
Comment (LukeMathWalker) [inline crates/polars-pipe/src/pipeline/dispatcher.rs:367]: I found this a bit strange—`sink_finished` is defined outside the "main" `for` loop, but here we only break the inner `while let` loop. Is that intentional?

**Comment by LukeMathWalker** [comment]
OK, something is not quite right—trying to figure it out, but I'm struggling.

File: crates/polars-lazy/src/tests/streaming.rs [hunk 2]
```diff
@@ -24,6 +24,7 @@ fn assert_streaming_with_default(q: LazyFrame, total: bool, check_shape_only: bo
     }
     let q_expected = q.with_streaming(false);
     let out = q_streaming.collect().unwrap();
+    dbg!("Non-streaming");
```
Comment (ritchie46) [inline crates/polars-lazy/src/tests/streaming.rs:27]: dbg!

File: crates/polars-pipe/src/pipeline/dispatcher.rs (hunk 1 above)
Comment (ritchie46) [reply crates/polars-pipe/src/pipeline/dispatcher.rs:367]: We should break the outer loop indeed.

**Review by ritchie46 — CHANGES_REQUESTED** [review]

**Comment by ritchie46** [comment]
Wow, this hasn't gotten any easier to follow. :')

Any idea what query caused the stackoverflow?

**Comment by LukeMathWalker** [comment]
The same query I mentioned in https://github.com/pola-rs/polars/issues/11829 minus the streaming part, with the important caveat that it is running over a large dataset that's partitioned into thousands of smaller files.

**Comment by ritchie46** [comment]
> The same query I mentioned in #11829 minus the streaming part, with the important caveat that it is running over a large Parquet dataset that's (Hive-)partitioned into thousands of small files.

Yes, I estimated that. This is now fixed by #11922. Where instead of a union per file, we create a single source that can handle all those files. So the stackoverflow should be resolved by better handling multi-file datasets. 

This currently is only done for parquet files, but I want to implement this for all file types we have. Can you give it a try with a compiled version of main?

**Comment by stinodego** [comment]
Since the original issue that this PR was trying address seems to have been resolved, I will close this. Thanks for the initiative though @LukeMathWalker !

**Pull request closed without merging**